### DISCUSSION TITLE
localizations cleanup

### DISCUSSION THREAD

**Discussion by rkitover** [visualboyadvance-m/visualboyadvance-m#1071] [General]
Let's discuss ongoing work reorganizing and cleaning up the translation support and translations here.

**Comment by rkitover** [comment]
@StanleyKid-22 So the other thing I have left to do here from that discussion is to mark that list of strings you gave me as not-to-be-translated?

**Reply by StanleyKid-22** [reply]
Yes, of course.
In this way, we'll at least approximately detect the true percentage of translated strings of each localization.
Of course, it would be better if the Poedit developer thought about the possibility of marking such strings, looks like formatting in msgid_str or other, (we already discussed in #1029 about that), and the problem would disappear by itself, and in Transifex we would see the real percentage. But if he doesn't respond to suggestions, which will help not only that, and also other projects where gettext is used - we just have no other way out.
And even if it suddenly turns out that some string of a certain localization, previously marked as not requiring translation, will need translate, it can be returned always, IMHO.

**Comment by Felipefpl** [comment]
I hope you seriously consider what i wrote here: https://github.com/visualboyadvance-m/visualboyadvance-m/issues/993 ;)

I really believe translations that arent at least 50% complete should be deleted. The brazilian portuguese translation was IIRC, 0% complete when i decided i'd translate everything. If ppl cared about that translation they'd finish the job like i did. ;)

**Comment by rkitover** [comment]
Hi, sorry I didn't have our discord bot wired to show discussions activity, I've fixed that.

**Comment by rkitover** [comment]
It looks like you can't reference discussions from commits.

@StanleyKid-22 Your latest set of source string fixes has been committed in d756f671.

I'll catch up more here later, by the way your posts are visible to all of us on the discord now, I fixed our event/webhook bot.

**Comment by Felipefpl** [comment]
Small changes - It would be nice if some strings were changed to ALWAYS use uppercase letters to avoid Poedit reporting problems that dont exist such as the ones below, all of them are related to that:

![Sem título](https://user-images.githubusercontent.com/5569804/218105053-a8f84e61-eb03-4b9c-a2d0-6e088e282f60.gif)

**Reply by StanleyKid-22** [reply]
I think these strings
```
#: panel.cpp:2444
msgid "memory allocation error"

#: panel.cpp:2447
msgid "error initializing codec"

#: panel.cpp:2450
msgid "error writing to output file"

#: panel.cpp:2453
msgid "can't guess output format from file name"

#: panel.cpp:2458
msgid "programming error; aborting!"
```

it's a part of another global string, so they begin with a lowercase, but I can be wrong.
As for `dsbSecondary` or `e-Reader` - there's peculiar text construction. The names of operations begin with a lowercase, 'cause this is a programming feature. It's possible, of course, to rearrange the text, to begin with a uppercase.
I don't know about `visualboyadvance-m` at all, so I left it without translation.

**Comment by StanleyKid-22** [comment]
By the way, wouldn't it be better to collect the translations in a separate folder (e. g. `Language`) so that they don't mix with other folders in the application root?

**Comment by StanleyKid-22** [comment]
I honestly tried to reach.
Looks like it's not meant to be. 😁 
https://github.com/vslavik/poedit/issues/793

**Reply by Felipefpl** [reply]
> I honestly tried to reach. Looks like it's not meant to be. grin [vslavik/poedit#793](https://github.com/vslavik/poedit/issues/793)

Dont worry, this happened with me too. Vaclav treated me the same way. ;)

**Comment by rkitover** [comment]
@StanleyKid-22 @Felipefpl @ZachBacon @Steelskin @Squall-Leonhart 

Hi everyone, it's been a year but we're finally getting back to this project. Sorry it took so long, but we've all had a very rough year, and I did as well.

Here's an update on what we did recently:

- @ZachBacon removed the very incomplete translations from transifex in 3ab048be.
- @StanleyKid-22 I finally marked the strings you wanted to remove from the translations catalog as no-translate in e70e57a5. I didn't know how to do this in the `xrc` source before but now I do.

. @Felipefpl you also wanted to clean up a few other translations on transifex, please post here what you wanted done and any other ideas you have.

Next steps:

- Redo the translation catalog loading code to load translations embedded into the Windows binary, but also allow downloading the `translations.zip` if someone wants to update their translations for a build. We also need to improve the translations loading on Linux and macOS, to allow using the `translations.zip` and to load translations from the build directory when doing dev work.
- Add a `Language` menu to choose the language and save it in the config. People have been asking for this for years now.

. I need to do my OpenGL refactor, and then I will get back to these two items and any others you have regarding translations.

Regarding the **OK** buttons, I think we need to change some of them to be more specific so that they are more translatable, e.g. **Continue**, **Save**, etc..

I also want to complete the Russian translation, but I forgot the Russian keyboard so that will be a challenge.

**Comment by Felipefpl** [comment]
> . @Felipefpl you also wanted to clean up a few other translations on transifex, please post here what you wanted done and any other ideas you have.

Thanks for the heads-up, here is what i have to say:

1 - Chinese has 2 systems: simplified and traditional (any other without this description should be deleted).

2 - It's pointless to have an english (US) translation file since there is the pot file from which we do our translations so it must be deleted too.
 
3 - There are 2 french/hungarian/korean/polish translations (one of them must go). ;)

4 - There are 5 spanish translations (only 2 should exist AFAIK: one for spain and another for latin america).

>     * Redo the translation catalog loading code to load translations embedded into the Windows binary, but also allow downloading the `translations.zip` if someone wants to update their translations for a build. We also need to improve the translations loading on Linux and macOS, to allow using the `translations.zip` and to load translations from the build directory when doing dev work.

Instead downloading a separate package for translations (vba-m is the ONLY emu where we have to do this) the translations should be embedded in the emulator like it happens with duckstation/pcsx2/every other emulator out there.

>     * Add a `Language` menu to choose the language and save it in the config. People have been asking for this for years now.
 
A good idea which i suggested too months ago.
 
> Regarding the **OK** buttons, I think we need to change some of them to be more specific so that they are more translatable, e.g. **Continue**, **Save**, etc..

Sorry but i disagree, everyone knows what OK is, this doesnt need a refactoring. Again, most of other emus out there uses OK. ;)

Edit - i forgot to ask again to change the strings that start with small letters that i reported above on Feb 10 so that poedit wont complain about it.

**Reply by StanleyKid-22** [reply]
> Instead downloading a separate package for translations (vba-m is the ONLY emu where we have to do this) the translations should be embedded in the emulator like it happens with duckstation/pcsx2/every other emulator out there.

If this talking about including the langpack to the EXE (especially encrypted without the ability to edit them in the resource editors), I disagree, 'cause it won't be possible to test the translation after hotfixes.
But a separate folder for translations wouldn't hurt, I paid attention to that earlier.

**Reply by ZachBacon** [reply]
> > Instead downloading a separate package for translations (vba-m is the ONLY emu where we have to do this) the translations should be embedded in the emulator like it happens with duckstation/pcsx2/every other emulator out there.
> 
> If this talking about including the langpack to the EXE (especially encrypted without the ability to edit them in the resource editors), I disagree, 'cause it won't be possible to test the translation after hotfixes.
> But a separate folder for translations wouldn't hurt, I paid attention to that earlier.

Considering you can't really edit the translations even in resource editors is really moot. Wxwidgets uses gettext which is itself it's own compiled translation blob.

**Comment by Felipefpl** [comment]
> If this talking about including the langpack to the EXE (especially encrypted without the ability to edit them in the resource editors), I disagree, 'cause it won't be possible to test the translation after hotfixes. But a separate folder for translations wouldn't hurt, I paid attention to that earlier.

It's a separate foler for translations like everyone does. ;)

**Reply by StanleyKid-22** [reply]
Well, OK.
Maybe I misunderstood the message.

**Comment by rkitover** [comment]
I am absolutely not including more folders in the distribution, the distribution will always be a single `.exe`. I want to include the translations in the `.exe` by default, but still allow reading translations from the directory it's in like it does now.

**Comment by rkitover** [comment]
@Felipefpl @StanleyKid-22 I fixed the strings that started with lowercase, there were only a few, sorry it took me so long to get to this.

**Comment by Felipefpl** [comment]
> @Felipefpl @StanleyKid-22 I fixed the strings that started with lowercase, there were only a few, sorry it took me so long to get to this.

Thank you for that. Btw, there is a 0% Urdu "translation" that can be removed. It was added but no one is translating anything.

**Reply by ZachBacon** [reply]
@Felipefpl Just took care of it

**Comment by StanleyKid-22** [comment]
Sorry about that, just an organizational question.
The localized file always states that I translated only until 2023, although I regularly update it until today. I noticed that this may be the case in other translations.
Is this a auto-generation glitch on Transifex, or supposed to be like that?..
